File: lintcode/reverse_linked_list/cpp/reverse_linked_list.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/reverse-linked-list
@Language: C++
@Datetime: 17-02-16 00:13
*/

/**
 * Definition of ListNode
 * 
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 * 
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: The new head of reversed linked list.
     */
    ListNode *reverse(ListNode *head) 
    {
        ListNode *prev{ nullptr },
                 *curr{ head };
                  
        while (curr != nullptr) 
        {
            auto next  = curr->next;
            curr->next = prev;
            prev       = curr;
            curr       = next;
        }
        
        return prev;
    }
};

File: lintcode/easy/reverse_linked_list/py/reverse_linked_list.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/reverse-linked-list
@Language: Python
@Datetime: 17-02-16 00:10
'''

"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: You should return the head of the reversed linked list. 
                  Reverse it in-place.
    """
    def reverse(self, head):
        prev = None
        curr = head
        
        while curr != None:
            temp      = curr.next
            curr.next = prev
            prev      = curr
            curr      = temp
            
        return prev